### PR TITLE
feat: Trash downloads folder before runs

### PR DESCRIPTION
**Pull request by chrisbreiding**
- Addresses [TR-602](https://cypress-io.atlassian.net/browse/TR-602)
- Addresses https://github.com/cypress-io/cypress/issues/949

### User facing changelog

_Not sure if this needs its own changelog item, as it's just a corollary to https://github.com/cypress-io/cypress/pull/14431. This might actually be considered a chore rather than a feature._

### Additional Details

With https://github.com/cypress-io/cypress/pull/14431, users can test out file downloads. With this, the downloads folder will be trashed along with screenshots and videos before a run unless `trashAssetsBeforeRuns` is `false`.

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has the original issue or this PR been tagged with a release in ZenHub? 
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/3472
- N/A Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- N/A Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

**Review by flotwig — COMMENTED** [review]
I wonder how many users will want to auto-trash `screenshots` and `videos`, but not `downloads`, or vice versa. We could extend `trashAssetsBeforeRuns` to accept `boolean | { screenshots?: boolean, videos?: boolean, downloads?: boolean }`.

But I guess that's probably not worth the effort, since the user can always add a plugin event to use `fs` to delete whatever they want between tests, and it's a rare use case.

Is there an existing docs PR for this? It looks like `trashAssetsBeforeRuns` docs will need an update: image

File: packages/server/lib/modes/run.js [hunk 1]
```diff
@@ -634,10 +634,11 @@ const trashAssets = Promise.method((config = {}) => {
     return
   }
 
-  return Promise.join(
+  return Promise.all([
     trash.folder(config.videosFolder),
     trash.folder(config.screenshotsFolder),
-  )
+    trash.folder(path.join(config.projectRoot, 'cypress', 'downloads')),
```
Comment (flotwig) [inline packages/server/lib/modes/run.js:640]: nice https://github.com/cypress-io/cypress/pull/14465 is gonna fix this up

File: packages/server/test/e2e/4_downloads_spec.ts [hunk 2]
```diff
@@ -1,14 +1,58 @@
...
+    await e2e.exec(this, {
+      project: downloadsProject,
+      spec: 'simple_passing_spec.ts',
+      config: {
+        trashAssetsBeforeRuns: false,
+      },
+    })
+
+    const filePath = path.join(downloadsProject, 'cypress', 'downloads', 'records.csv')
+    const exists = await fs.pathExists(filePath)
```
Comment (flotwig) [inline packages/server/test/e2e/4_downloads_spec.ts:54]: is there anywhere that the contents of the download are asserted on? maybe worth adding a test if not

Comment (chrisbreiding) [reply packages/server/test/e2e/4_downloads_spec.ts:54]: They're asserted on in the integration test itself.

Comment (flotwig) [reply packages/server/test/e2e/4_downloads_spec.ts:54]: oh i see now https://github.com/cypress-io/cypress/blob/TR-602-clear-downloads/packages/server/test/support/fixtures/projects/downloads/cypress/integration/downloads_spec.ts

**Comment by chrisbreiding** [comment]
> I wonder how many users will want to auto-trash screenshots and videos, but not downloads, or vice versa. We could extend trashAssetsBeforeRuns to accept boolean | { screenshots?: boolean, videos?: boolean, downloads?: boolean }.

> But I guess that's probably not worth the effort, since the user can always add a plugin event to use fs to delete whatever they want between tests, and it's a rare use case.

We discussed this and decided we might change it later if users report issues with it not being broken up.

> Is there an existing docs PR for this? It looks like trashAssetsBeforeRuns docs will need an update: 

Yeah, it's part of https://github.com/cypress-io/cypress-documentation/pull/3472. I'll add it to the PR description.

**Review by flotwig — CHANGES_REQUESTED** [review]
Good to go once updated to use `config.downloadsFolder`

**Review by flotwig — DISMISSED** [review]
looks great :)

**Pull request merged**